### DISCUSSION TITLE
Smart method decision for small difference in return types

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/csharplang#5820] [Language Ideas] [answered]
Since :
![image](https://user-images.githubusercontent.com/35312807/155314423-5e2d4bd1-6356-455a-b73b-704ad839f78c.png)
These two statements works the same, perhaps I can expect .NET to be smart enough to not return anything when the value is discarded at call site?

so :
```
_=Mutate(new Func<LDS,LDS)((e)=>{});
```
Is equivalent to calling 
![image](https://user-images.githubusercontent.com/35312807/155314660-4e4e3ece-23e9-4f4f-a3b9-26c62e387c0c.png)
and :
```
var MyStruct =Mutate(new Func<LDS,LDS)((e)=>{});
```
Is equivalent to calling 
![image](https://user-images.githubusercontent.com/35312807/155314788-a82d3a33-2d91-42b7-945d-bf626413df50.png)
Therefore I can just write one method instead of two, having its function signature as to be returning LDS

**Comment by svick** [comment, marked as answer]
> Therefore I can just write one method instead of two, having its function signature as to be returning LDS

Is there is a specific reason why you can't do that today? At least based on my limited test, [`Mutate(...);` and `_ = Mutated(...);` generate the same machine code](https://godbolt.org/z/r76643sPW):

```asm
C:Mutated(byref):this:
       push     rbx
       mov      rbx, rsi
       mov      rax, gword ptr [rdi+8]
       mov      esi, dword ptr [rbx]
       mov      rdi, gword ptr [rax+8]
       call     qword ptr [rax+24]System.Func`2:Invoke(LDS):LDS:this
       mov      dword ptr [rbx], eax
       pop      rbx
       ret      

C:Mutate(byref):this:
       push     rbx
       mov      rbx, rsi
       mov      rax, gword ptr [rdi+8]
       mov      esi, dword ptr [rbx]
       mov      rdi, gword ptr [rax+8]
       call     qword ptr [rax+24]System.Func`2:Invoke(LDS):LDS:this
       mov      dword ptr [rbx], eax
       pop      rbx
       ret   
```

**Reply by Xyncgas** [reply]
Interesting

**Comment by HaloFour** [comment]
The IL for the two are necessarily different.  One leaves an `LDS` on the stack, and the other leaves nothing on the stack.  But as @svick pointed out that difference can be eliminated by JIT.  Or you can eliminate the one method and just ignore the return value if you prefer.  That's pretty much exactly what the machine code is doing.
